cmd/login.go

```diff
@@ -45,45 +59,59 @@ func login(_ *cobra.Command, _ []string) {
 	}
 
 	groupName := ospreyconfig.GroupOrDefault(targetGroup)
-	snapshot := client.GetSnapshot(ospreyconfig)
+	snapshot := ospreyconfig.GetSnapshot()
 	group, ok := snapshot.GetGroup(groupName)
 	if !ok {
 		log.Errorf("Group not found: %q", groupName)
 		os.Exit(1)
 	}
 
-	credentials, err := client.GetCredentials()
-	if err != nil {
-		log.Fatalf("Failed to get credentials: %v", err)
-	}
-
 	displayActiveGroup(targetGroup, ospreyconfig.DefaultGroup)
+	retreiverOptions := client.RetreiverOptions{
+		UseDeviceCode: useDeviceCode,
+		LoginTimeout:  loginTimeout,
+	}
 
 	success := true
-	for _, target := range group.Targets() {
-		c := client.NewClient(target.Server(), ospreyconfig.CertificateAuthorityData, target.CertificateAuthorityData())
-		tokenData, err := c.GetAccessToken(credentials)
+
+	retrieverFactory, err := client.NewProviderFactory(ospreyconfig, retreiverOptions)
+	if err != nil {
+		log.Errorf("Unable to initialise providers: %v", err)
+	}
+
+	for provider, targets := range group.Targets() {
+		retriever, err := retrieverFactory.GetRetriever(provider)
 		if err != nil {
-			if state, ok := status.FromError(err); ok && state.Code() == codes.Unauthenticated {
-				log.Fatalf("Failed to log in to %s: %v", target.Name(), state.Message())
+			log.Fatalf("Unsupported provider %s: %v", provider, err.Error())
+		}
+		for _, target := range targets {
+			targetData, err := retriever.RetrieveClusterDetailsAndAuthTokens(target)
+			if err != nil {
+				if state, ok := status.FromError(err); ok && state.Code() == codes.Unauthenticated {
+					log.Fatalf("Failed to log in to %s: %v", target.Name(), state.Message())
+				}
+				success = false
+				log.Errorf("Failed to log in to %s: %v", target.Name(), err)
+				continue
 			}
-			success = false
-			log.Errorf("Failed to log in to %s: %v", target.Name(), err)
-			continue
+			updateKubeconfig(target, targetData)
 		}
+	}
 
-		err = kubeconfig.UpdateConfig(target.Name(), target.Aliases(), tokenData)
-		if err != nil {
-			log.Errorf("Failed to update config for %s: %v", target.Name(), err)
-		}
+	if !success {
+		log.Fatal("Failed to update credentials for some targets.")
+	}
+}
+
+func updateKubeconfig(target client.Target, tokenData *client.TargetInfo) {
+	err := kubeconfig.UpdateConfig(target.Name(), target.Aliases(), tokenData)
+	if err != nil {
+		log.Errorf("Failed to update config for %s: %v", target.Name(), err)
```

Comment:
```suggestion
		log.Errorf("Failed to update config for %s: %v", target.Name(), err)
		return
```
Can we add a return there, and remove the `else` branch bellow?